java-engine/src/main/java/io/getunleash/engine/UnleashEngine.java

```diff
@@ -99,55 +97,70 @@ public VariantDef getVariant(String name, Context context)
                     new TypeReference<YggResponse<VariantDef>>() {
                     });
-            return response.getValue();
+            return response;
         } catch (JsonProcessingException e) {
             throw new YggdrasilInvalidInputException(context);
         }
     }
 
     public void countToggle(String flagName, boolean enabled) {
-        yggdrasil.countToggle(this.enginePtr, flagName, enabled);
+        yggdrasil.countToggle(this.enginePtr, toUtf8Pointer(flagName), enabled);
     }
 
     public void countVariant(String flagName, String variantName) {
-        yggdrasil.countVariant(this.enginePtr, flagName, variantName);
+        yggdrasil.countVariant(this.enginePtr, toUtf8Pointer(flagName),
+                toUtf8Pointer(variantName));
     }
 
     public MetricsBucket getMetrics() throws YggdrasilError {
-        YggResponse<MetricsBucket> response = read(yggdrasil.getMetrics(this.enginePtr),
+        MetricsBucket response = read(yggdrasil.getMetrics(this.enginePtr),
                 new TypeReference<YggResponse<MetricsBucket>>() {
                 });
-        return response.getValue();
+        return response;
     }
 
     public boolean shouldEmitImpressionEvent(String name) throws YggdrasilError {
-        YggResponse<Boolean> response = read(
-                yggdrasil.shouldEmitImpressionEvent(this.enginePtr, name),
+        Boolean response = read(
+                yggdrasil.shouldEmitImpressionEvent(this.enginePtr, toUtf8Pointer(name)),
                 new TypeReference<YggResponse<Boolean>>() {
                 });
-        return response.getValue();
+        return response;
     }
 
     public List<FeatureDef> listKnownToggles() throws YggdrasilError {
-        YggResponse<List<FeatureDef>> response = read(
+        List<FeatureDef> response = read(
                 yggdrasil.listKnownToggles(this.enginePtr),
                 new TypeReference<YggResponse<List<FeatureDef>>>() {
                 });
-        return response.getValue();
+        return response;
     }
 
     public static String getCoreVersion() {
         Pointer versionPointer = UnleashFFI.getYggdrasilCoreVersion();
         return versionPointer.getString(0);
     }
 
+    /**
+     * Handle reading from a pointer into an YggdrasilResponse and unwrapping that
+     * to an object
+     */
+    private <T> T read(Pointer pointer, TypeReference<YggResponse<T>> clazz) throws YggdrasilError {
+        YggResponse<T> wrappedResponse = readRaw(pointer, clazz);
+        if (wrappedResponse.isValid()) {
+            return wrappedResponse.getValue();
+        } else {
+            throw new YggdrasilError(wrappedResponse.errorMessage);
+        }
+    }
+
     /** Handle reading from a pointer into a String and mapping it to an object */
-    private <T> T read(Pointer pointer, TypeReference<T> clazz) {
+    private <T> T readRaw(Pointer pointer, TypeReference<T> clazz) {
```

Comment:
Maintained for cases where Yggdrasil returns static entities that are not encoded with the normal response formatting (JSON blob signalling if the call succeeded etc)